File: python/examples/quickstart.py
from __future__ import annotations

import os

import psycopg
from psycopg_pool import ConnectionPool

from workhorse import Queue, Worker

database_url = os.environ["DATABASE_URL"]

with psycopg.connect(database_url) as application_connection:
    task_id = Queue(application_connection).enqueue("email.welcome", {"to": "ada@example.com"})
    application_connection.commit()

with ConnectionPool(
    database_url, min_size=3, max_size=3, kwargs={"autocommit": True}
) as worker_pool:
    worker = Worker(worker_pool).handle(
        "email.welcome",
        lambda payload, _context: {"deliveredTo": payload["to"]},
    )
    assert worker.run_once() is True  # Production worker processes call run().

print(task_id)

File: rust/examples/quickstart.rs
//! Enqueues one task, runs the worker once, and prints the task's outcome.
// docs:start quickstart-program
use serde_json::{json, Value};
use workhorse::deadpool_postgres::tokio_postgres::NoTls;
use workhorse::deadpool_postgres::{Manager, Pool};
use workhorse::{Admin, EnqueueOptions, Queue, Worker, WorkerOptions};

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    let url = std::env::var("DATABASE_URL")?;
    let queue = Queue::connect(&url, "default").await?;
    let enqueued = queue
        .enqueue("email.welcome", &json!({ "to": "ada@example.com" }), EnqueueOptions::default())
        .await?;

    let pool = Pool::builder(Manager::new(url.parse()?, NoTls)).max_size(4).build()?;
    let worker = Worker::new(pool, WorkerOptions::default())?;
    worker.handle("email.welcome", |payload: Value, _context| async move {
        Ok(json!({ "deliveredTo": payload["to"] }))
    });
    worker.run_once().await?; // production uses run_worker_process(&worker)

    let admin = Admin::connect(&url).await?;
    if let Some(task) = admin.get_task(enqueued.task_id).await? {
        println!("{:?} {}", task.state, task.result.unwrap_or_default());
    }
    Ok(())
}
// docs:end
